core/types/calls.cc

```diff
@@ -2175,6 +2175,30 @@ class Magic_selfNew : public IntrinsicMethod {
     }
 } Magic_selfNew;
 
+class Magic_splat : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, const DispatchArgs &args, DispatchResult &res) const override {
+        if (args.args.size() != 1) {
+            return;
+        }
+
+        auto &arg = args.args[0];
+        if (auto *tuple = cast_type<TupleType>(arg->type)) {
+            res.returnType = arg->type;
+            return;
+        }
+
+        if (auto *ap = cast_type<AppliedType>(arg->type)) {
+            if (ap->klass == Symbols::Array() || ap->klass.data(gs)->derivesFrom(gs, Symbols::Array())) {
+                res.returnType = arg->type;
+                return;
+            }
+        }
+
+        res.returnType = Types::untypedUntracked();
```

Comment:
I'm not sure that this intrinsic is returning the correct type. Here's an example from irb:
```
irb(main):001:0> xs = [1,2,3]
=> [1, 2, 3]
irb(main):002:0> ys = *xs
=> [1, 2, 3]
irb(main):003:0> ys.class
=> Array
irb(main):004:0> ys = *1
=> [1]
irb(main):005:0> ys.class
=> Array
```

Reply:
Hm, I guess the right thing to do here would be:

* Attempt to dispatch `arg.to_a`
* If that returns something, use that
* If that returns an error (or something non-arrayish?), then try to return a one-element `Array` (maybe untyped, depending?)

?  I tried a quick and dumb version of this and returned `Array[T.untyped]`, but that ran afoul of code in `callWithSplat` that was expecting `untyped` or `TupleType`, and I was not clever enough to see what introducing arrays there might do.  I guess if `callWithSplat` gets an array type, it should just treat that as untyped?